command/agent/http.go

```diff
@@ -75,89 +76,111 @@ type HTTPServer struct {
 	wsUpgrader *websocket.Upgrader
 }
 
-// NewHTTPServer starts new HTTP server over the agent
-func NewHTTPServer(agent *Agent, config *Config) (*HTTPServer, error) {
-	// Start the listener
-	lnAddr, err := net.ResolveTCPAddr("tcp", config.normalizedAddrs.HTTP)
-	if err != nil {
-		return nil, err
-	}
-	ln, err := config.Listener("tcp", lnAddr.IP.String(), lnAddr.Port)
-	if err != nil {
-		return nil, fmt.Errorf("failed to start HTTP listener: %v", err)
-	}
+// NewHTTPServers starts an HTTP server for every address.http configured in
+// the agent.
+func NewHTTPServers(agent *Agent, config *Config) ([]HTTPServer, error) {
+	var srvs []HTTPServer
+	var error error
 
-	// If TLS is enabled, wrap the listener with a TLS listener
-	if config.TLSConfig.EnableHTTP {
-		tlsConf, err := tlsutil.NewTLSConfiguration(config.TLSConfig, config.TLSConfig.VerifyHTTPSClient, true)
+	// Start the listener
+	for _, addr := range config.normalizedAddrs.HTTP {
+		lnAddr, err := net.ResolveTCPAddr("tcp", addr)
 		if err != nil {
-			return nil, err
+			multierror.Append(error, err)
+			continue
 		}
-
-		tlsConfig, err := tlsConf.IncomingTLSConfig()
+		ln, err := config.Listener("tcp", lnAddr.IP.String(), lnAddr.Port)
 		if err != nil {
-			return nil, err
+			error = multierror.Append(error, fmt.Errorf("failed to start HTTP listener: %v", err))
+			continue
 		}
-		ln = tls.NewListener(tcpKeepAliveListener{ln.(*net.TCPListener)}, tlsConfig)
-	}
 
-	// Create the mux
-	mux := http.NewServeMux()
+		// If TLS is enabled, wrap the listener with a TLS listener
+		if config.TLSConfig.EnableHTTP {
+			tlsConf, err := tlsutil.NewTLSConfiguration(config.TLSConfig, config.TLSConfig.VerifyHTTPSClient, true)
+			if err != nil {
+				error = multierror.Append(error, err)
+				continue
+			}
 
-	wsUpgrader := &websocket.Upgrader{
-		ReadBufferSize:  2048,
-		WriteBufferSize: 2048,
-	}
+			tlsConfig, err := tlsConf.IncomingTLSConfig()
+			if err != nil {
+				error = multierror.Append(error, err)
+				continue
+			}
+			ln = tls.NewListener(tcpKeepAliveListener{ln.(*net.TCPListener)}, tlsConfig)
+		}
 
-	// Create the server
-	srv := &HTTPServer{
-		agent:      agent,
-		mux:        mux,
-		listener:   ln,
-		listenerCh: make(chan struct{}),
-		logger:     agent.httpLogger,
-		Addr:       ln.Addr().String(),
-		wsUpgrader: wsUpgrader,
-	}
-	srv.registerHandlers(config.EnableDebug)
+		// Create the mux
+		mux := http.NewServeMux()
```

Comment:
Can the mux be reused across listeners?

Reply:
spend a couple minutes investigating this. Its possible, but a lot of the functions that initialize the handlers are predicated on a single server.

If you move the mux out of the loop, you get 
```
==> Starting Nomad agent...                                                                                                                            [56/315]
panic: http: multiple registrations for /v1/jobs

goroutine 1 [running]:
net/http.(*ServeMux).Handle(0xc000878380, {0x245e573, 0x8}, {0x2937000, 0xc0007f42d0})
        net/http/server.go:2441 +0x226
net/http.(*ServeMux).HandleFunc(...)
        net/http/server.go:2478
github.com/hashicorp/nomad/command/agent.(*HTTPServer).registerHandlers(0xc000667310, 0x0)
        github.com/hashicorp/nomad/command/agent/http.go:271 +0x96
github.com/hashicorp/nomad/command/agent.NewHTTPServers(0xc000293680, 0xc0007ece00)
        github.com/hashicorp/nomad/command/agent/http.go:154 +0x7e7
github.com/hashicorp/nomad/command/agent.(*Command).setupAgent(0xc000786180, 0xc0007ece00, {0x29cda40, 0xc000855b60}, {0x2934820, 0xc0007f5320}, 0x14)
        github.com/hashicorp/nomad/command/agent/command.go:503 +0xed
github.com/hashicorp/nomad/command/agent.(*Command).Run(0xc000786180, {0xc00004c1a0, 0x1, 0x1})
        github.com/hashicorp/nomad/command/agent/command.go:693 +0x5d8
github.com/mitchellh/cli.(*CLI).Run(0xc00076af00)
        github.com/mitchellh/cli@v1.1.0/cli.go:260 +0x5f8
main.RunCustom({0xc00004c190, 0x1e4fdc0, 0xc0000001a0})
        github.com/hashicorp/nomad/main.go:116 +0x33d
main.Run(...)
        github.com/hashicorp/nomad/main.go:86
main.main()
        github.com/hashicorp/nomad/main.go:82 +0x50
```
because the registerHandlers function (https://github.com/hashicorp/nomad/blob/147d97145b3c3ad0edde6b852fe90e0e4243c813/command/agent/http.go#L251-L371) is being called for each server. I investigated if we can make it a 'regular function' instead of being implemented for the `HTTPServer` struct, but it would require turning all of the route handlers and the wrap function into 'regular functions' as well which may not be worth doing in this PR.

Does that make sense? Is that the only way of doing this? Do we want to try and break that out after this is done?

Reply:
Refactoring the whole mux doesn't seem like a great idea just in terms of effort. Seems fine to leave this where it is for now and try to get as much of the other components out of the loop as we can.